NullPointerException on Config load

Hello,

I'm running into a bit of an issue with classes whose hashcode implementation relies on the Configuration, but who are themselves part of the Configuration.

For example, I have a ValueClass that wraps a String with some validation, which I use - among other places - in my configuration. However, the hashcode implementation of that class calls `.lowercase(MyConfig.configuredLocale)` on the string, to make sure strings are treated case-insensitively.
Now this crashes on load, because Hoplite calls the hashcode method on the object during config loading ( `com.sksamuel.hoplite.NodeState.hashCode(DecoderContext.kt)` ), and finds that the CONFIG does not exist yet and throws a runtime NPE (on startup) because of it.

I can fix this inside the hashcode method by doing something like `.lowercase(MyConfig?.configuredLocale ?: Locale.US)`, but this is highlighted by the IDE as bad because MyConfig is not nullable and thus "can never be null".
Alternatively I can abstract the configuration in some method that returns a nullable config, but then I have to deal with nullability and call chains on nested properties `MyConfig?.look?.like?.this?.and?.the?.question?.marks?.are?.making?.me?.question?.my ?: life`

This all feels bad to just work around the case that some functions are called during config initialisation on a value that regularly is guaranteed non-null. Do you have any ideas on how I can do this cleanly?

Here is the most minimal example I could come up with:

```kotlin
object MyDecoder: Decoder<MyClass> {
  override fun decode(node: Node, type: KType, context: DecoderContext): ConfigResult<MyClass> =
    when(node) {
      is StringNode -> MyClass(node.value).valid()
      else -> ConfigFailure.DecodeError(node, type).invalid()
    }

  override fun supports(type: KType): Boolean = type.isSubtypeOf(MyClass::class.starProjectedType)
}

val CONFIG = ConfigLoaderBuilder.default()
  .addResourceSource("/application.yml")
  .addDecoder(MyDecoder)
  .build()
  .loadConfigOrThrow<MyConfig>()

data class MyConfig(
  val someInt: Int,
  val test: MyClass,
)

class MyClass(val myString: String) {
  override fun hashCode(): Int =
    myString.hashCode() + CONFIG.someInt
}

fun main() {
  println(CONFIG)
}

```

Error is:
```stacktrace
Exception in thread "main" java.lang.ExceptionInInitializerError
Caused by: java.lang.NullPointerException: Cannot invoke "MyConfig.getSomeInt()" because the return value of "MyConfigKt.getCONFIG()" is null
	at MyClass.hashCode(MyConfig.kt:32)
	at com.sksamuel.hoplite.NodeState.hashCode(DecoderContext.kt)
	at java.base/java.util.HashMap.hash(HashMap.java:338)
        ...
```

--- a/hoplite-core/src/main/kotlin/com/sksamuel/hoplite/DecoderContext.kt
+++ b/hoplite-core/src/main/kotlin/com/sksamuel/hoplite/DecoderContext.kt
@@ -25,7 +25,7 @@ data class DecoderContext(
   // these are the dot paths for every config value - overrided or not, that was used
   val usedPaths: MutableSet<DotPath> = mutableSetOf(),
   // this tracks the types that a node was marshalled into
-  val used: MutableSet<NodeState> = mutableSetOf(),
+  val used: MutableMap<DotPath, NodeState> = mutableMapOf(),
   val metadata: MutableMap<String, Any?> = mutableMapOf(),
   val config: DecoderConfig = DecoderConfig(false),
   val environment: Environment? = null,
@@ -49,7 +49,7 @@ data class DecoderContext(
    * Makes a node as marshalled into the given [type] with the resolved value [value].
    */
   fun used(node: Node, type: KType, value: Any?) {
-    this.used.add(NodeState(node, true, value, type))
+    this.used[node.path] = NodeState(node, true, value, type)
   }
 
   /**

--- a/hoplite-core/src/main/kotlin/com/sksamuel/hoplite/internal/Decoding.kt
+++ b/hoplite-core/src/main/kotlin/com/sksamuel/hoplite/internal/Decoding.kt
@@ -30,7 +30,7 @@ class Decoding(
   }
 }
 
-fun createDecodingState(
+internal fun createDecodingState(
   root: Node,
   context: DecoderContext,
   secretsPolicy: SecretsPolicy?
@@ -48,7 +48,7 @@ private fun createNodeStates(
 ): List<NodeState> {
   return root.traverse().map { node ->
 
-    val state = context.used.find { it.node.path == node.path }
+    val state = context.used.entries.find { it.key == node.path }?.value
 
     val secret = secretsPolicy?.isSecret(node, state?.type) ?: false
 

--- a/hoplite-core/src/test/kotlin/com/sksamuel/hoplite/Github389.kt
+++ b/hoplite-core/src/test/kotlin/com/sksamuel/hoplite/Github389.kt
@@ -1,0 +1,47 @@
+package com.sksamuel.hoplite
+
+import com.sksamuel.hoplite.decoder.Decoder
+import com.sksamuel.hoplite.fp.invalid
+import com.sksamuel.hoplite.fp.valid
+import com.sksamuel.hoplite.parsers.PropsPropertySource
+import io.kotest.core.spec.style.FunSpec
+import java.util.Properties
+import kotlin.reflect.KType
+import kotlin.reflect.full.isSubtypeOf
+import kotlin.reflect.full.starProjectedType
+
+// https://github.com/sksamuel/hoplite/issues/389
+class Github389 : FunSpec() {
+  init {
+    test("NullPointerException on Config load #389") {
+      val props = Properties()
+      props["someInt"] = 1
+      props["test"] = "a"
+      ConfigLoaderBuilder.default()
+        .addPropertySource(PropsPropertySource(props))
+        .addDecoder(MyDecoder)
+        .build()
+        .loadConfigOrThrow<MyConfig>()
+    }
+  }
+
+}
+
+object MyDecoder : Decoder<MyClass> {
+  override fun decode(node: Node, type: KType, context: DecoderContext): ConfigResult<MyClass> =
+    when (node) {
+      is StringNode -> MyClass(node.value).valid()
+      else -> ConfigFailure.DecodeError(node, type).invalid()
+    }
+
+  override fun supports(type: KType): Boolean = type.isSubtypeOf(MyClass::class.starProjectedType)
+}
+
+class MyClass(val value: String) {
+  override fun hashCode(): Int = error("do not run at config loadtime")
+}
+
+data class MyConfig(
+  val someInt: Int,
+  val test: MyClass,
+)

